src/wallet/wallet.h

```diff
@@ -789,6 +789,10 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
 
     ~CWallet()
     {
+        // Even if RegisterValidationInstance was never called on this instance,
+        // the unregister call is safe to make and is a no-op.
+        UnregisterValidationInterface(this);
```

Comment:
I think this doesn't work because UnregisterValidationInterface tries to call virtual methods, which isn't something you can do in c++ in the middle of object destruction. I ran into a similar problem in #10973

https://github.com/bitcoin/bitcoin/blob/0bdf0e330d5d16e8e6a93d7d4ad663db812c4c91/src/interface/chain.cpp#L163-L169

Reply:
Hmm, so then should I revert to the version where the Register call was moved to the bottom of the method?

Reply:
> Hmm, so then should I revert to the version where the Register call was moved to the bottom of the method?

IIRC the problem with moving the registration is that it creates potential for race conditions if the code is called at any time other than startup, which is something that's been attempted in prs like #10740 (and I also want to support in #10102).

I think I'd recommend just fixing the memory leak with the minimal suggestion from https://github.com/bitcoin/bitcoin/pull/12647#discussion_r174875670. But if you prefer moving the registration, that also seems fine with a comment mentioning the potential for race conditions.

I think maybe the best way to clean up error handling and leaks here going forward would be to move initialization logic out of the static `CWallet::CreateWalletFromFile` method into a non-static `bool CWallet::Initialize(...) method`. This way errors could just be handled by returning `false` from `Initialize` and `CreateWalletFromFile` would only be responsible for constructing the wallet, calling Initialize, and deleting/unregistering the wallet object in the case of errors.

Reply:
Is there actually a potential for race conditions? The function locks cs_main before the rescan logic and before the `RegisterValidationInterface` call even before it was moved.